poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+*
+ *   2) PromiseState = responsible for keeping state about a given Promise such
+ *                     as shares recovered and messages either certifying the
+ *                     promise or proving that it is malicious
+ *
+ *   3) PromiseSignature = proves that an insurer has signed off on a Promise.
+ *                         The signature could either be used to express
+ *                         approval or disapproval
+ *
+ *   4) BlameProof = provides proof that a given Promise share was malicously
+ *                   constructed. A valid BlameProof proves that the Promise
+ *                   is untrustworthy and that the creator of the Promise is
+ *                   malicious
+ *
+ * Further documentation for each of the different structs can be found below.
+ * It is suggested to start with the Promise struct referring to the others
+ * as necessary. Once a general knowledge of Promise is gained, the others
+ * will make more sense.
+ *
+ * Code using this package will typically have the following flow (please see
+ * "Key Terms" below for a definition of terms used):
+ *
+ * 1) The promiser constructs a new Promise and stores it withing a PromiseState
+ *
+ * 2) The promiser sends the Promise to the insurers
+ *
+ * 3) The insurers verify the Promise is well-formed and make sure that their
+ *    secret shares are valid.
+ *
+ * 4) If the secret share is invalid, an insurer creates a BlameProof and sends
+ *    it back.
+ *
+ * 5) If the share is valid, an insurer creates a PromiseSignature to send back
+ *    to the promiser.
+ *
+ * 6) The promiser receives the message from the insurer.
+ *
+ * 7) If it is a valid BlameProof, the promiser must start all over and
+ *    construct a non-malicious Promise (or, the system can ban this malicious
+ *    promiser).
+ *
+ * 8) If the message is a PromiseSignature, the promiser can add the signature
+ *    to its PromiseState
+ *
+ * 9) Steps 2-8 are repeated until the promiser has collected enough
+ *    PromiseSignatures for the Promise to be considered certified.
+ *
+ * 10) Once the Promise is certified, the promiser can then send the promise to
+ *     clients.
+ *
+ * 11) Clients can then request the signatures from the insurers to make sure
+ *     the Promise is indeed certified
+ *
+ *      i) This prevents a malicious promiser from simply leaving out valid
+ *         BlameProofs and only sending good signatures to the clients.
+ *
+ * 12) The promiser can perform some work for clients.
+ *
+ * 13) If the promiser is unresponsive for too long, a client can inform the
+ *     insurers of the Promise.
+ *
+ * 14) The insurers can then check if the promiser is indeed unresponsive.
+ *
+ * 15) If so, an insurer reveal its share and send it to the client.
+ *
+ * 16) The client repeats steps 13-15 until enough shares are recovered to
+ *     reconstruct the secret.
+ *
+ * 17) The client can then reconstruct the secret and take over for the promiser
+ *
+ *
```

Comment:
In general, you need more structure above: any list of 17 steps is by definition pretty much impenetrable to a reader.